Catch all errors from commands and log them rather than returning them

- Add a trait of the form:

    ```rust
    trait ForAll {
        fn accept(&self, project: &Project) -> bool {
            true
        }

        fn run(&mut self, project: &Project, opts: Options) -> Result<(), ForAllError>;

        // Used by cloc (along with the `&mut` above):
        fn report(&self) -> std::io::Result<()> {
            Ok(())
        }
    }
    ```

- Give each command a method that returns `Box<dyn ForAll>`, and have
  `commands::Command::run()` execute this object as follows:

    ```rust
    let mut failures = Vec::new();
    for p in projects {
        if forall.accept(&p) {
            if let Err(e) = forall.run(&p) {
                log::error!( todo!("something about e") );
                if opts.keep_going {
                    failures.push(p);
                } else {
                    return ExitCode::FAILURE;
                }
            }
        }
    }
    forall.report()?;  // TODO: Handle error
    if failures.is_empty() {
        ExitCode::SUCCESS
    } else {
        todo!("Print failures");
        ExitCode::FAILURE
    }
    ```

- `ForAllError` should be an enum of:
    - Command errors
        - Display includes command line, return code, and any captured output
    - HTTP errors
        - Status errors are displayed prettily
    - Custom messages
    - IO errors?
    - ???

- The `run()` methods are responsible for calling `logproject()`

- idea: Display error chains in the form:

    ```
    [ERROR] Top-level error
    ⤷ Caused by: source 0
    ⤷ Caused by: source 1
    ⤷ Caused by: source 2
    ```

- Indent HTTP error responses and failed command output somehow

- Be sure to update the "Logging" section of the README to note that error messages include output from quieted commands

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ v0.3.0 (in development)
 - Added `--verbose` option
 - `--quiet` can now be specified twice to suppress output from `run` and
   `runpr` commands
+- `--keep-going` now captures all error types
 
 v0.2.0 (2025-01-06)
 -------------------

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -24,7 +24,6 @@ mime = "0.3.17"
 parse_link_header = { version = "0.4.0", features = ["url"] }
 rand = "0.9.0"
 serde = { version = "1.0.215", features = ["derive"] }
-serde-jsonlines = "0.7.0"
 serde_json = "1.0.133"
 shell-words = "1.1.0"
 thiserror = "2.0.6"

--- a/README.md
+++ b/README.md
@@ -51,10 +51,10 @@ subcommand.
   in the given language.  Possible options are "Python"/"py" and "Rust"/"rs"
   (all case-insensitive).
 
-- `-k`, `--keep-going` — By default, if a subcommand fails for a project,
-  `forall` terminates immediately.  If `--keep-going` is supplied, `forall`
-  will instead continue with the remaining projects and will print a list of
-  all failures on exit.
+- `-k`, `--keep-going` — By default, if a subcommand fails or another error
+  occurs for a project, `forall` terminates immediately.  If `--keep-going` is
+  supplied, `forall` will instead continue with the remaining projects and will
+  print a list of all failures on exit.
 
 - `-q`, `--quiet` — Be less verbose; this option can be specified multiple
   times.  See "Logging" below for more infomation.
@@ -81,6 +81,7 @@ shown for each quiet/verbose level:
 | --------------------------------- | :---: | :--: | :-: | :--: | ------ | ------ |
 | Project names                     | ✓     | ✓    | ✓   | ✓    | Bold   | stdout |
 | Errors                            | ✓     | ✓    | ✓   | ✓    | Red    | stderr |
+| Lists of failures                 | ✓     | ✓    | ✓   | ✓    | Plain  | stdout |
 | `run` and `runpr` commands        | ✗     | ✗    | ✓   | ✓    | Cyan   | stderr |
 | `run` and `runpr` commands output | ✗     | ✓    | ✓   | ✓    | Plain  | stdout |
 | Operational commands              | ✗     | ✗    | ✓   | ✓    | Cyan   | stderr |
@@ -98,6 +99,9 @@ Notes:
 
 - Passing more than two `--quiet` options (after applying `--verbose` negation)
   is equivalent to passing just two.
+
+- "Errors" includes captured output from failed commands whose output would
+  otherwise be suppressed.
 
 - The "commands" message types are messages showing each executed command along
   with its arguments and working directory.

--- a/src/cmd.rs
+++ b/src/cmd.rs
@@ -37,7 +37,6 @@ pub(crate) struct CommandPlus {
     cmdline: CommandLine,
     cmd: Command,
     kind: CommandKind,
-    keep_going: bool,
     stderr_set: bool,
 }
 
@@ -48,7 +47,6 @@ impl CommandPlus {
             cmdline: CommandLine::new(arg0),
             cmd: Command::new(arg0),
             kind: CommandKind::default(),
-            keep_going: false,
             stderr_set: false,
         }
     }
@@ -96,16 +94,11 @@ impl CommandPlus {
         self
     }
 
-    pub(crate) fn keep_going(&mut self, yes: bool) -> &mut Self {
-        self.keep_going = yes;
-        self
-    }
-
     pub(crate) fn cmdline(&self) -> &CommandLine {
         &self.cmdline
     }
 
-    pub(crate) fn run(&mut self) -> Result<bool, CommandError> {
+    pub(crate) fn run(&mut self) -> Result<(), CommandError> {
         logcmd(self, self.kind.cmdline_verbosity());
         let rc = if !is_active(self.kind.output_verbosity()) {
             let (output, rc) = self.combine_stdout_stderr()?;
@@ -121,14 +114,7 @@ impl CommandPlus {
             })?
         };
         if rc.success() {
-            Ok(true)
-        } else if self.keep_going {
-            if let Some(code) = rc.code() {
-                error!("[{code}]");
-            } else {
-                error!("[{rc}]");
-            }
-            Ok(false)
+            Ok(())
         } else {
             Err(CommandError::Exit {
                 cmdline: self.cmdline().clone(),

--- a/src/cmd.rs
+++ b/src/cmd.rs
@@ -131,7 +131,7 @@ impl CommandPlus {
         })
     }
 
-    pub(crate) fn check_output(&mut self) -> Result<String, CommandOutputError> {
+    pub(crate) fn check_output(&mut self) -> Result<String, CommandError> {
         logcmd(self, self.kind.cmdline_verbosity());
         if !self.stderr_set {
             self.cmd.stderr(Stdio::inherit());
@@ -139,23 +139,23 @@ impl CommandPlus {
         match self.cmd.output() {
             Ok(output) if output.status.success() => match String::from_utf8(output.stdout) {
                 Ok(s) => Ok(s),
-                Err(e) => Err(CommandOutputError::Decode {
+                Err(e) => Err(CommandError::Decode {
                     cmdline: self.cmdline().clone(),
                     source: e.utf8_error(),
                 }),
             },
-            Ok(output) => Err(CommandOutputError::Exit {
+            Ok(output) => Err(CommandError::Exit {
                 cmdline: self.cmdline().clone(),
                 rc: output.status,
             }),
-            Err(source) => Err(CommandOutputError::Startup {
+            Err(source) => Err(CommandError::Startup {
                 cmdline: self.cmdline().clone(),
                 source,
             }),
         }
     }
 
-    fn combine_stdout_stderr(&mut self) -> Result<(String, ExitStatus), CombinedCommandError> {
+    fn combine_stdout_stderr(&mut self) -> Result<(String, ExitStatus), CommandError> {
         logcmd(self, self.kind.cmdline_verbosity());
         // <https://stackoverflow.com/a/72831067/744178>
         let mut child = self
@@ -163,7 +163,7 @@ impl CommandPlus {
             .stdout(Stdio::piped())
             .stderr(Stdio::piped())
             .spawn()
-            .map_err(|source| CombinedCommandError::Startup {
+            .map_err(|source| CommandError::Startup {
                 cmdline: self.cmdline().clone(),
                 source,
             })?;
@@ -210,7 +210,7 @@ impl CommandPlus {
 
         drop(sender);
 
-        let rc = child.wait().map_err(|source| CombinedCommandError::Wait {
+        let rc = child.wait().map_err(|source| CommandError::Wait {
             cmdline: self.cmdline().clone(),
             source,
         })?;
@@ -218,7 +218,7 @@ impl CommandPlus {
         match stdout_thread.join() {
             Ok(Ok(())) => (),
             Ok(Err(source)) => {
-                return Err(CombinedCommandError::Read {
+                return Err(CommandError::Read {
                     cmdline: self.cmdline().clone(),
                     source,
                 })
@@ -229,7 +229,7 @@ impl CommandPlus {
         match stderr_thread.join() {
             Ok(Ok(())) => (),
             Ok(Err(source)) => {
-                return Err(CombinedCommandError::Read {
+                return Err(CommandError::Read {
                     cmdline: self.cmdline().clone(),
                     source,
                 })
@@ -293,35 +293,10 @@ pub(crate) enum CommandError {
         cmdline: CommandLine,
         rc: ExitStatus,
     },
-    #[error(transparent)]
-    Combined(#[from] CombinedCommandError),
-}
-
-#[derive(Debug, Error)]
-pub(crate) enum CommandOutputError {
-    #[error("failed to run {cmdline:#}")]
-    Startup {
-        cmdline: CommandLine,
-        source: std::io::Error,
-    },
-    #[error("command {cmdline:#} failed: {rc}")]
-    Exit {
-        cmdline: CommandLine,
-        rc: ExitStatus,
-    },
     #[error("could not decode {cmdline:#} output")]
     Decode {
         cmdline: CommandLine,
         source: std::str::Utf8Error,
-    },
-}
-
-#[derive(Debug, Error)]
-pub(crate) enum CombinedCommandError {
-    #[error("failed to run {cmdline:#}")]
-    Startup {
-        cmdline: CommandLine,
-        source: std::io::Error,
     },
     #[error("error reading from {cmdline:#}")]
     Read {

--- a/src/cmd.rs
+++ b/src/cmd.rs
@@ -1,7 +1,7 @@
 use crate::logging::{is_active, logcmd, Verbosity};
 use std::ffi::OsStr;
 use std::fmt;
-use std::io::{BufRead, BufReader, Write};
+use std::io::{BufRead, BufReader};
 use std::path::{Path, PathBuf};
 use std::process::{Command, ExitStatus, Stdio};
 use thiserror::Error;
@@ -100,18 +100,17 @@ impl CommandPlus {
 
     pub(crate) fn run(&mut self) -> Result<(), CommandError> {
         logcmd(self, self.kind.cmdline_verbosity());
-        let rc = if !is_active(self.kind.output_verbosity()) {
+        let (output, rc) = if !is_active(self.kind.output_verbosity()) {
             let (output, rc) = self.combine_stdout_stderr()?;
-            if !rc.success() {
-                // TODO: Better error handling here:
-                let _ = std::io::stdout().write_all(output.as_bytes());
-            }
-            rc
+            (Some(output), rc)
         } else {
-            self.cmd.status().map_err(|source| CommandError::Startup {
-                cmdline: self.cmdline().clone(),
-                source,
-            })?
+            (
+                None,
+                self.cmd.status().map_err(|source| CommandError::Startup {
+                    cmdline: self.cmdline().clone(),
+                    source,
+                })?,
+            )
         };
         if rc.success() {
             Ok(())
@@ -119,6 +118,7 @@ impl CommandPlus {
             Err(CommandError::Exit {
                 cmdline: self.cmdline().clone(),
                 rc,
+                output,
             })
         }
     }
@@ -147,6 +147,7 @@ impl CommandPlus {
             Ok(output) => Err(CommandError::Exit {
                 cmdline: self.cmdline().clone(),
                 rc: output.status,
+                output: String::from_utf8(output.stdout).ok(),
             }),
             Err(source) => Err(CommandError::Startup {
                 cmdline: self.cmdline().clone(),
@@ -292,6 +293,7 @@ pub(crate) enum CommandError {
     Exit {
         cmdline: CommandLine,
         rc: ExitStatus,
+        output: Option<String>,
     },
     #[error("could not decode {cmdline:#} output")]
     Decode {

--- a/src/cmd.rs
+++ b/src/cmd.rs
@@ -312,6 +312,16 @@ pub(crate) enum CommandError {
     },
 }
 
+impl CommandError {
+    pub(crate) fn output(&self) -> Option<&str> {
+        if let CommandError::Exit { output, .. } = self {
+            output.as_deref()
+        } else {
+            None
+        }
+    }
+}
+
 fn quote_osstr(s: &OsStr) -> String {
     shell_words::quote(&s.to_string_lossy()).to_string()
 }

--- a/src/commands/clean.rs
+++ b/src/commands/clean.rs
@@ -1,21 +1,19 @@
+use super::ForAll;
 use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::Options;
 use clap::Args;
 
 /// Run `git clean -dXf` on each project
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct Clean;
 
-impl Clean {
-    pub(crate) fn run(self, _opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        for p in projects {
-            if !p.readcmd("git", ["clean", "-dXn"])?.is_empty() {
-                logproject(&p);
-                p.runcmd("git").args(["clean", "-dXf"]).run()?;
-            } else {
-                debug!("{}: already clean", p.name());
-            }
+impl ForAll for Clean {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        if !p.readcmd("git", ["clean", "-dXn"])?.is_empty() {
+            logproject(p);
+            p.runcmd("git").args(["clean", "-dXf"]).run()?;
+        } else {
+            debug!("{}: already clean", p.name());
         }
         Ok(())
     }

--- a/src/commands/cloc.rs
+++ b/src/commands/cloc.rs
@@ -1,49 +1,29 @@
-use crate::cmd::CommandOutputError;
-use crate::logging::logfailures;
+use super::ForAll;
 use crate::project::{Language, Project};
-use crate::util::Options;
 use anyhow::Context;
 use clap::Args;
 use serde::Deserialize;
 
 /// Count effective lines of code in each project
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct Cloc;
 
-impl Cloc {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        let mut failures = Vec::new();
-        for p in projects {
-            let srcs = p.source_paths()?;
-            if srcs.is_empty() {
-                if opts.keep_going {
-                    error!("{}: Could not identify source files", p.name());
-                    failures.push(p);
-                    continue;
-                } else {
-                    anyhow::bail!("{}: Could not identify source files", p.name());
-                }
-            }
-            let r = p
-                .runcmd("cloc")
-                .arg(format!("--include-ext={}", p.language().ext()))
-                .arg("--json")
-                .args(srcs)
-                .check_output();
-            let output = match r {
-                Ok(output) => output,
-                Err(CommandOutputError::Exit { .. }) if opts.keep_going => {
-                    failures.push(p);
-                    continue;
-                }
-                Err(e) => return Err(e.into()),
-            };
-            let data = serde_json::from_str::<ClocJson>(&output)
-                .context("failed to deserialize `cloc` output")?;
-            let lines = data.for_language(p.language()).unwrap_or_default().code;
-            println!("{lines:6} {}", p.name());
+impl ForAll for Cloc {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        let srcs = p.source_paths()?;
+        if srcs.is_empty() {
+            anyhow::bail!("{}: Could not identify source files", p.name());
         }
-        logfailures(failures);
+        let output = p
+            .runcmd("cloc")
+            .arg(format!("--include-ext={}", p.language().ext()))
+            .arg("--json")
+            .args(srcs)
+            .check_output()?;
+        let data = serde_json::from_str::<ClocJson>(&output)
+            .context("failed to deserialize `cloc` output")?;
+        let lines = data.for_language(p.language()).unwrap_or_default().code;
+        println!("{lines:6} {}", p.name());
         Ok(())
     }
 }

--- a/src/commands/gc.rs
+++ b/src/commands/gc.rs
@@ -1,18 +1,16 @@
+use super::ForAll;
 use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::Options;
 use clap::Args;
 
 /// Run `git gc` on each project
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct Gc;
 
-impl Gc {
-    pub(crate) fn run(self, _opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        for p in projects {
-            logproject(&p);
-            p.runcmd("git").arg("gc").run()?;
-        }
+impl ForAll for Gc {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        logproject(p);
+        p.runcmd("git").arg("gc").run()?;
         Ok(())
     }
 }

--- a/src/commands/list.rs
+++ b/src/commands/list.rs
@@ -1,27 +1,24 @@
+use super::ForAll;
 use crate::project::Project;
-use crate::util::Options;
 use clap::Args;
-use serde_jsonlines::JsonLinesWriter;
 
 /// List all projects
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct List {
     /// Output JSON
     #[arg(short = 'J', long)]
     json: bool,
 }
 
-impl List {
-    pub(crate) fn run(self, _opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
+impl ForAll for List {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
         if self.json {
-            let mut out = JsonLinesWriter::new(std::io::stdout());
-            for p in projects {
-                out.write(&p.to_details()?)?;
-            }
+            println!(
+                "{}",
+                serde_json::to_string(&p.to_details()?).expect("JSONification should not fail")
+            );
         } else {
-            for p in projects {
-                println!("{}", p.name());
-            }
+            println!("{}", p.name());
         }
         Ok(())
     }

--- a/src/commands/mod.rs
+++ b/src/commands/mod.rs
@@ -18,6 +18,7 @@ use self::push::Push;
 use self::rsclean::Rsclean;
 pub(crate) use self::run::Run;
 use self::runpr::RunPr;
+use crate::logging::logerror;
 use crate::project::Project;
 use crate::util::Options;
 use clap::Subcommand;
@@ -55,16 +56,14 @@ impl Command {
             Command::Run(c) => match c.into_forall() {
                 Ok(cmd) => cmd,
                 Err(e) => {
-                    // TODO: Display more stuff:
-                    error!("Failed to initialize command: {e}");
+                    logerror(e.context("Failed to initialize command"));
                     return ExitCode::FAILURE;
                 }
             },
             Command::RunPr(c) => match c.into_forall() {
                 Ok(cmd) => cmd,
                 Err(e) => {
-                    // TODO: Display more stuff:
-                    error!("Failed to initialize command: {e}");
+                    logerror(e.context("Failed to initialize command"));
                     return ExitCode::FAILURE;
                 }
             },
@@ -72,7 +71,7 @@ impl Command {
         let mut failures = Vec::new();
         for p in projects {
             if let Err(e) = cmd.run(&p) {
-                error!("{e}"); // TODO: Improve; display causes, output, & HTTP responses
+                logerror(e);
                 if opts.keep_going {
                     failures.push(p);
                 } else {

--- a/src/commands/mod.rs
+++ b/src/commands/mod.rs
@@ -21,6 +21,11 @@ use self::runpr::RunPr;
 use crate::project::Project;
 use crate::util::Options;
 use clap::Subcommand;
+use std::process::ExitCode;
+
+trait ForAll {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()>;
+}
 
 #[derive(Clone, Debug, Eq, PartialEq, Subcommand)]
 pub(crate) enum Command {
@@ -37,18 +42,55 @@ pub(crate) enum Command {
 }
 
 impl Command {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        match self {
-            Command::List(c) => c.run(opts, projects),
-            Command::Clean(c) => c.run(opts, projects),
-            Command::Cloc(c) => c.run(opts, projects),
-            Command::Gc(c) => c.run(opts, projects),
-            Command::PreUpdate(c) => c.run(opts, projects),
-            Command::Pull(c) => c.run(opts, projects),
-            Command::Push(c) => c.run(opts, projects),
-            Command::Rsclean(c) => c.run(opts, projects),
-            Command::Run(c) => c.run(opts, projects),
-            Command::RunPr(c) => c.run(opts, projects),
+    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> ExitCode {
+        let mut cmd: Box<dyn ForAll> = match self {
+            Command::List(c) => Box::new(c),
+            Command::Clean(c) => Box::new(c),
+            Command::Cloc(c) => Box::new(c),
+            Command::Gc(c) => Box::new(c),
+            Command::PreUpdate(c) => Box::new(c),
+            Command::Pull(c) => Box::new(c),
+            Command::Push(c) => Box::new(c),
+            Command::Rsclean(c) => Box::new(c),
+            Command::Run(c) => match c.into_forall() {
+                Ok(cmd) => cmd,
+                Err(e) => {
+                    // TODO: Display more stuff:
+                    error!("Failed to initialize command: {e}");
+                    return ExitCode::FAILURE;
+                }
+            },
+            Command::RunPr(c) => match c.into_forall() {
+                Ok(cmd) => cmd,
+                Err(e) => {
+                    // TODO: Display more stuff:
+                    error!("Failed to initialize command: {e}");
+                    return ExitCode::FAILURE;
+                }
+            },
+        };
+        let mut failures = Vec::new();
+        for p in projects {
+            if let Err(e) = cmd.run(&p) {
+                error!("{e}"); // TODO: Improve; display causes, output, & HTTP responses
+                if opts.keep_going {
+                    failures.push(p);
+                } else {
+                    return ExitCode::FAILURE;
+                }
+            }
+        }
+        if failures.is_empty() {
+            ExitCode::SUCCESS
+        } else {
+            anstream::println!(
+                "\n{bold}Failures:{bold:#}",
+                bold = anstyle::Style::new().bold()
+            );
+            for p in failures {
+                println!("{}", p.name());
+            }
+            ExitCode::FAILURE
         }
     }
 }

--- a/src/commands/preupdate.rs
+++ b/src/commands/preupdate.rs
@@ -1,67 +1,45 @@
+use super::ForAll;
 use crate::cmd::CommandError;
-use crate::logging::{logfailures, logproject};
+use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::Options;
 use clap::Args;
 use fs_err::PathExt;
 
 static PRE_COMMIT_FILE: &str = ".pre-commit-config.yaml";
 
 /// Update .pre-commit-config.yaml files
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct PreUpdate;
 
-impl PreUpdate {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        let mut failures = Vec::new();
-        for p in projects {
-            if !p.dirpath().join(PRE_COMMIT_FILE).fs_err_try_exists()? {
-                debug!("{} does not use pre-commit; skipping", p.name());
-                continue;
-            }
-            logproject(&p);
-            p.stash()?;
-            if !p
-                .runcmd("pre-commit")
-                .arg("autoupdate")
-                .keep_going(opts.keep_going)
-                .run()?
-            {
-                failures.push(p);
-                continue;
-            }
-            p.runcmd("git").args(["add", PRE_COMMIT_FILE]).run()?;
-            // TODO: Suppress the "[{returncode}]" output when this fails:
-            p.runcmd("pre-commit")
-                .args(["run", "-a"])
-                .keep_going(true)
-                .run()?;
-            p.runcmd("git").args(["add", "-a"]).run()?;
-            // Run pre-commit again in order to check for linting & similar
-            // errors without the rewriting of files (which also causes
-            // pre-commit to exit nonzero) causing false positives:
-            if !p
-                .runcmd("pre-commit")
-                .args(["run", "-a"])
-                .keep_going(opts.keep_going)
-                .run()?
-            {
-                failures.push(p);
-                continue;
-            }
-            match p.runcmd("git").args(["diff", "--cached", "--quiet"]).run() {
-                Ok(_) => (),
-                Err(CommandError::Exit { .. }) => {
-                    p.runcmd("git")
-                        .args(["commit", "-m"])
-                        .arg(format!("Autoupdate {PRE_COMMIT_FILE}"))
-                        .keep_going(opts.keep_going)
-                        .run()?;
-                }
-                Err(e) => return Err(e.into()),
-            }
+impl ForAll for PreUpdate {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        if !p.dirpath().join(PRE_COMMIT_FILE).fs_err_try_exists()? {
+            debug!("{} does not use pre-commit; skipping", p.name());
+            return Ok(());
         }
-        logfailures(failures);
+        logproject(p);
+        p.stash()?;
+        p.runcmd("pre-commit").arg("autoupdate").run()?;
+        p.runcmd("git").args(["add", PRE_COMMIT_FILE]).run()?;
+        match p.runcmd("pre-commit").args(["run", "-a"]).run() {
+            Ok(()) | Err(CommandError::Exit { .. }) => (),
+            Err(e) => return Err(e.into()),
+        }
+        p.runcmd("git").args(["add", "-a"]).run()?;
+        // Run pre-commit again in order to check for linting & similar
+        // errors without the rewriting of files (which also causes
+        // pre-commit to exit nonzero) causing false positives:
+        p.runcmd("pre-commit").args(["run", "-a"]).run()?;
+        match p.runcmd("git").args(["diff", "--cached", "--quiet"]).run() {
+            Ok(()) => (),
+            Err(CommandError::Exit { .. }) => {
+                p.runcmd("git")
+                    .args(["commit", "-m"])
+                    .arg(format!("Autoupdate {PRE_COMMIT_FILE}"))
+                    .run()?;
+            }
+            Err(e) => return Err(e.into()),
+        }
         Ok(())
     }
 }

--- a/src/commands/pull.rs
+++ b/src/commands/pull.rs
@@ -1,33 +1,22 @@
-use crate::logging::{logfailures, logproject};
+use super::ForAll;
+use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::Options;
 use clap::Args;
 
 /// Run `git pull` on each project
 ///
 /// Only projects that have GitHub remotes are considered.
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct Pull;
 
-impl Pull {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        let mut failures = Vec::new();
-        for p in projects {
-            if !p.has_github() {
-                debug!("{} does not have a GitHub repository; skipping", p.name());
-                continue;
-            }
-            logproject(&p);
-            if !p
-                .runcmd("git")
-                .arg("pull")
-                .keep_going(opts.keep_going)
-                .run()?
-            {
-                failures.push(p);
-            }
+impl ForAll for Pull {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        if !p.has_github() {
+            debug!("{} does not have a GitHub repository; skipping", p.name());
+        } else {
+            logproject(p);
+            p.runcmd("git").arg("pull").run()?;
         }
-        logfailures(failures);
         Ok(())
     }
 }

--- a/src/commands/push.rs
+++ b/src/commands/push.rs
@@ -1,42 +1,29 @@
-use crate::logging::{logfailures, logproject};
+use super::ForAll;
+use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::Options;
 use clap::Args;
 
 /// Run `git push` on each project
 ///
 /// Only projects that have GitHub remotes and for which `HEAD` is ahead of
 /// `@{upstream}` are considered.
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct Push;
 
-impl Push {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        let mut failures = Vec::new();
-        for p in projects {
-            if !p.has_github() {
-                debug!("{} does not have a GitHub repository; skipping", p.name());
-                continue;
-            }
-            // TODO: If this fails, emit "{BOLD:name}\n{ERROR:[1]}" and handle
-            // with keep_going:
+impl ForAll for Push {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        if !p.has_github() {
+            debug!("{} does not have a GitHub repository; skipping", p.name());
+        } else {
             let ahead = p.readcmd(
                 "git",
                 ["rev-list", "--count", "--right-only", "@{upstream}...HEAD"],
             )?;
             if ahead.parse::<usize>().unwrap_or_default() > 0 {
-                logproject(&p);
-                if !p
-                    .runcmd("git")
-                    .arg("push")
-                    .keep_going(opts.keep_going)
-                    .run()?
-                {
-                    failures.push(p);
-                }
+                logproject(p);
+                p.runcmd("git").arg("push").run()?;
             }
         }
-        logfailures(failures);
         Ok(())
     }
 }

--- a/src/commands/rsclean.rs
+++ b/src/commands/rsclean.rs
@@ -1,35 +1,24 @@
-use crate::logging::{logfailures, logproject};
+use super::ForAll;
+use crate::logging::logproject;
 use crate::project::{Language, Project};
-use crate::util::Options;
 use clap::Args;
 use fs_err::PathExt;
 
 /// Run `cargo clean` on Rust projects with `target/` directories
-#[derive(Args, Clone, Debug, Eq, PartialEq)]
+#[derive(Args, Clone, Copy, Debug, Eq, PartialEq)]
 pub(crate) struct Rsclean;
 
-impl Rsclean {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
-        let mut failures = Vec::new();
-        for p in projects {
-            if p.language() != Language::Rust || !p.dirpath().join("target").fs_err_try_exists()? {
-                debug!(
-                    "{} is not a Rust project with a target/ directory; skipping",
-                    p.name()
-                );
-                continue;
-            }
-            logproject(&p);
-            if !p
-                .runcmd("cargo")
-                .arg("clean")
-                .keep_going(opts.keep_going)
-                .run()?
-            {
-                failures.push(p);
-            }
+impl ForAll for Rsclean {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        if p.language() != Language::Rust || !p.dirpath().join("target").fs_err_try_exists()? {
+            debug!(
+                "{} is not a Rust project with a target/ directory; skipping",
+                p.name()
+            );
+        } else {
+            logproject(p);
+            p.runcmd("cargo").arg("clean").run()?;
         }
-        logfailures(failures);
         Ok(())
     }
 }

--- a/src/commands/run.rs
+++ b/src/commands/run.rs
@@ -1,6 +1,7 @@
-use crate::logging::{logfailures, logproject};
+use super::ForAll;
+use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::{Options, RunOpts, Runner};
+use crate::util::{RunOpts, Runner};
 use clap::Args;
 
 /// Run a command on each project.
@@ -18,19 +19,28 @@ pub(crate) struct Run {
 }
 
 impl Run {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
+    pub(super) fn into_forall(self) -> anyhow::Result<Box<dyn ForAll>> {
         let runner = Runner::try_from(self.run_opts)?;
-        let mut failures = Vec::new();
-        for p in projects {
-            logproject(&p);
-            if self.stash {
-                p.stash()?;
-            }
-            if !runner.run(&p, opts)? {
-                failures.push(p);
-            }
+        Ok(Box::new(RunForAll {
+            runner,
+            stash: self.stash,
+        }))
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct RunForAll {
+    runner: Runner,
+    stash: bool,
+}
+
+impl ForAll for RunForAll {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        logproject(p);
+        if self.stash {
+            p.stash()?;
         }
-        logfailures(failures);
+        self.runner.run(p)?;
         Ok(())
     }
 }

--- a/src/commands/runpr.rs
+++ b/src/commands/runpr.rs
@@ -1,7 +1,8 @@
+use super::ForAll;
 use crate::github::{CreateLabel, CreatePullRequest, GitHub};
-use crate::logging::{logfailures, logproject};
+use crate::logging::logproject;
 use crate::project::Project;
-use crate::util::{Options, RunOpts, Runner};
+use crate::util::{RunOpts, Runner};
 use clap::Args;
 use rand::{rng, seq::IndexedRandom, Rng};
 use std::borrow::Cow;
@@ -63,9 +64,9 @@ pub(crate) struct RunPr {
 }
 
 impl RunPr {
-    pub(crate) fn run(self, opts: Options, projects: Vec<Project>) -> anyhow::Result<()> {
+    pub(super) fn into_forall(self) -> anyhow::Result<Box<dyn ForAll>> {
         let github = GitHub::authed()?;
-        let mut colorgen = RandomColor::new(rng());
+        let colorgen = RandomColor::new(rng());
         let branch = match self.branch {
             Some(b) => b,
             None => OffsetDateTime::now_local()
@@ -76,95 +77,122 @@ impl RunPr {
         let pr_title = self
             .pr_title
             .as_deref()
-            .unwrap_or_else(|| strip_skip(&self.message));
+            .unwrap_or_else(|| strip_skip(&self.message))
+            .to_owned();
         let pr_body = match self.pr_body_file {
             Some(p) => Some(fs_err::read_to_string(p)?),
             None => None,
         };
         let runner = Runner::try_from(self.run_opts)?;
-        let mut failures = Vec::new();
-        for p in projects {
-            let Some(ghrepo) = p.ghrepo() else {
-                debug!("{} does not have a GitHub repository; skipping", p.name());
-                continue;
-            };
-            if github.get_repository(ghrepo)?.archived {
-                debug!("Repository for {} is archived; skipping", p.name());
-                continue;
-            }
-            logproject(&p);
-            let defbranch = p.default_branch()?;
-            p.stash()?;
+        Ok(Box::new(RunPrForAll {
+            github,
+            colorgen,
+            branch,
+            pr_title,
+            pr_body,
+            runner,
+            label: self.label,
+            soft_label: self.soft_label,
+            message: self.message,
+        }))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct RunPrForAll {
+    github: GitHub,
+    colorgen: RandomColor<rand::rngs::ThreadRng>,
+    branch: String,
+    pr_title: String,
+    pr_body: Option<String>,
+    runner: Runner,
+    label: Vec<String>,
+    soft_label: Vec<String>,
+    message: String,
+}
+
+impl ForAll for RunPrForAll {
+    fn run(&mut self, p: &Project) -> anyhow::Result<()> {
+        let Some(ghrepo) = p.ghrepo() else {
+            debug!("{} does not have a GitHub repository; skipping", p.name());
+            return Ok(());
+        };
+        if self.github.get_repository(ghrepo)?.archived {
+            debug!("Repository for {} is archived; skipping", p.name());
+            return Ok(());
+        }
+        logproject(p);
+        let defbranch = p.default_branch()?;
+        p.stash()?;
+        p.runcmd("git")
+            .arg("checkout")
+            .arg("-b")
+            .arg(&self.branch)
+            .arg(defbranch)
+            .run()?;
+        self.runner.run(p)?;
+        p.runcmd("git").args(["add", "."]).run()?;
+        // XXX: When adding support for commands that commit, also check
+        //      whether $branch is ahead of $defbranch.
+        if !p.has_staged_changes()? {
+            info!("No changes");
+            p.runcmd("git").arg("checkout").arg(defbranch).run()?;
             p.runcmd("git")
-                .arg("checkout")
-                .arg("-b")
-                .arg(&branch)
-                .arg(defbranch)
+                .args(["branch", "-d"])
+                .arg(&self.branch)
                 .run()?;
-            if !runner.run(&p, opts)? {
-                failures.push(p);
-                continue;
+            return Ok(());
+        }
+        p.runcmd("git")
+            .args(["commit", "-m"])
+            .arg(&self.message)
+            .run()?;
+        p.runcmd("git")
+            .args(["push", "--set-upstream", "origin"])
+            .arg(&self.branch)
+            .run()?;
+        let pr = self.github.create_pull_request(
+            ghrepo,
+            CreatePullRequest {
+                title: Cow::from(&self.pr_title),
+                head: Cow::from(&self.branch),
+                base: Cow::from(defbranch),
+                body: self.pr_body.as_deref().map(Cow::from),
+                maintainer_can_modify: true,
+            },
+        )?;
+        println!("{}", pr.html_url); // TODO: Improve display?
+        if !self.label.is_empty() || !self.soft_label.is_empty() {
+            let label_names = self
+                .github
+                .get_label_names(ghrepo)?
+                .into_iter()
+                .map(|s| s.to_ascii_lowercase())
+                .collect::<HashSet<_>>();
+            let mut labels = Vec::new();
+            for lbl in &self.label {
+                if !label_names.contains(&lbl.to_ascii_lowercase()) {
+                    self.github.create_label(
+                        ghrepo,
+                        CreateLabel {
+                            name: Cow::from(lbl),
+                            color: Cow::from(self.colorgen.generate()),
+                            description: None,
+                        },
+                    )?;
+                    info!("Created label {lbl:?} in {ghrepo}");
+                }
+                labels.push(lbl.as_str());
             }
-            p.runcmd("git").args(["add", "."]).run()?;
-            // XXX: When adding support for commands that commit, also check
-            //      whether $branch is ahead of $defbranch.
-            if !p.has_staged_changes()? {
-                info!("No changes");
-                p.runcmd("git").arg("checkout").arg(defbranch).run()?;
-                p.runcmd("git").args(["branch", "-d"]).arg(&branch).run()?;
-                continue;
-            }
-            p.runcmd("git")
-                .args(["commit", "-m"])
-                .arg(&self.message)
-                .run()?;
-            p.runcmd("git")
-                .args(["push", "--set-upstream", "origin"])
-                .arg(&branch)
-                .run()?;
-            let pr = github.create_pull_request(
-                ghrepo,
-                CreatePullRequest {
-                    title: Cow::from(pr_title),
-                    head: Cow::from(&branch),
-                    base: Cow::from(defbranch),
-                    body: pr_body.as_deref().map(Cow::from),
-                    maintainer_can_modify: true,
-                },
-            )?;
-            println!("{}", pr.html_url); // TODO: Improve display?
-            if !self.label.is_empty() || !self.soft_label.is_empty() {
-                let label_names = github
-                    .get_label_names(ghrepo)?
-                    .into_iter()
-                    .map(|s| s.to_ascii_lowercase())
-                    .collect::<HashSet<_>>();
-                let mut labels = Vec::new();
-                for lbl in &self.label {
-                    if !label_names.contains(&lbl.to_ascii_lowercase()) {
-                        github.create_label(
-                            ghrepo,
-                            CreateLabel {
-                                name: Cow::from(lbl),
-                                color: Cow::from(colorgen.generate()),
-                                description: None,
-                            },
-                        )?;
-                        info!("Created label {lbl:?} in {ghrepo}");
-                    }
+            for lbl in &self.soft_label {
+                if label_names.contains(&lbl.to_ascii_lowercase()) {
                     labels.push(lbl.as_str());
                 }
-                for lbl in &self.soft_label {
-                    if label_names.contains(&lbl.to_ascii_lowercase()) {
-                        labels.push(lbl.as_str());
-                    }
-                }
-                if !labels.is_empty() {
-                    github.add_labels_to_pr(ghrepo, pr.number, &labels)?;
-                }
+            }
+            if !labels.is_empty() {
+                self.github.add_labels_to_pr(ghrepo, pr.number, &labels)?;
             }
         }
-        logfailures(failures);
         Ok(())
     }
 }

--- a/src/http_util.rs
+++ b/src/http_util.rs
@@ -1,12 +1,12 @@
-use indenter::indented;
 use mime::{Mime, JSON};
 use serde_json::{to_string_pretty, value::Value};
-use std::fmt::{self, Write};
+use thiserror::Error;
 use ureq::Response;
 use url::Url;
 
 /// Error raised for a 4xx or 5xx HTTP response that includes the response body
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Eq, Error, PartialEq)]
+#[error("{method} request to {url} returned {status}")]
 pub(crate) struct StatusError {
     url: String,
     method: String,
@@ -33,23 +33,11 @@ impl StatusError {
             method: method.to_string(),
         }
     }
-}
 
-impl fmt::Display for StatusError {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(
-            f,
-            "{} request to {} returned {}",
-            self.method, self.url, self.status
-        )?;
-        if let Some(text) = &self.body {
-            write!(indented(f).with_str("    "), "\n\n{text}\n")?;
-        }
-        Ok(())
+    pub(crate) fn body(&self) -> Option<&str> {
+        self.body.as_deref()
     }
 }
-
-impl std::error::Error for StatusError {}
 
 /// Returns `true` iff the response's Content-Type header indicates the body is
 /// JSON

--- a/src/logging.rs
+++ b/src/logging.rs
@@ -1,13 +1,15 @@
-use crate::cmd::CommandPlus;
+use crate::cmd::{CommandError, CommandPlus};
 use crate::project::Project;
 use anstyle::{AnsiColor, Style};
+use indenter::indented;
+use std::fmt::{self, Write};
 use std::sync::OnceLock;
 
 static VERBOSITY: OnceLock<Verbosity> = OnceLock::new();
 
 #[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
 pub(crate) enum Verbosity {
-    On,
+    //On,
     Quiet2,
     Quiet,
     #[default]
@@ -52,16 +54,6 @@ pub(crate) fn logrequest(method: &str, url: &url::Url) {
     }
 }
 
-macro_rules! error {
-    ($($arg:tt)*) => {{
-        $crate::logging::logln(
-            $crate::logging::Verbosity::On,
-            anstyle::Style::new().fg_color(Some(anstyle::AnsiColor::Red.into())),
-            format_args!($($arg)*)
-        );
-    }};
-}
-
 macro_rules! info {
     ($($arg:tt)*) => {{
         $crate::logging::logln(
@@ -82,8 +74,33 @@ macro_rules! debug {
     }};
 }
 
-pub(crate) fn logln(level: Verbosity, style: Style, fmtargs: std::fmt::Arguments<'_>) {
+pub(crate) fn logln(level: Verbosity, style: Style, fmtargs: fmt::Arguments<'_>) {
     if is_active(level) {
         anstream::eprintln!("{style}[·] {fmtargs}{style:#}");
+    }
+}
+
+pub(crate) fn logerror(e: anyhow::Error) {
+    let style = Style::new().fg_color(Some(AnsiColor::BrightRed.into()));
+    anstream::eprintln!("{style}[!] {e}{style:#}");
+    for src in e.chain().skip(1) {
+        anstream::eprintln!("{style}[!] ⤷ {src}{style:#}");
+    }
+    if let Some(output) = e
+        .downcast_ref::<CommandError>()
+        .and_then(|src| src.output())
+    {
+        if !output.is_empty() {
+            anstream::eprint!("{style}{text}{style:#}", text = IndentOutput(output));
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct IndentOutput<'a>(&'a str);
+
+impl fmt::Display for IndentOutput<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(indented(f).with_str("[Output] "), "{}", self.0)
     }
 }

--- a/src/logging.rs
+++ b/src/logging.rs
@@ -1,4 +1,5 @@
 use crate::cmd::{CommandError, CommandPlus};
+use crate::http_util::StatusError;
 use crate::project::Project;
 use anstyle::{AnsiColor, Style};
 use indenter::indented;
@@ -91,16 +92,26 @@ pub(crate) fn logerror(e: anyhow::Error) {
         .and_then(|src| src.output())
     {
         if !output.is_empty() {
-            anstream::eprint!("{style}{text}{style:#}", text = IndentOutput(output));
+            anstream::eprint!(
+                "{style}{text}{style:#}",
+                text = Indented(output, "[Output] ")
+            );
+        }
+    } else if let Some(body) = e.downcast_ref::<StatusError>().and_then(|src| src.body()) {
+        if !body.is_empty() {
+            anstream::eprint!(
+                "{style}{text}{style:#}",
+                text = Indented(body, "[Response] ")
+            );
         }
     }
 }
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
-struct IndentOutput<'a>(&'a str);
+struct Indented<'a>(&'a str, &'static str);
 
-impl fmt::Display for IndentOutput<'_> {
+impl fmt::Display for Indented<'_> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(indented(f).with_str("[Output] "), "{}", self.0)
+        write!(indented(f).with_str(self.1), "{}", self.0)
     }
 }

--- a/src/logging.rs
+++ b/src/logging.rs
@@ -52,15 +52,6 @@ pub(crate) fn logrequest(method: &str, url: &url::Url) {
     }
 }
 
-pub(crate) fn logfailures(failures: Vec<Project>) {
-    if !failures.is_empty() {
-        anstream::println!("\n{bold}Failures:{bold:#}", bold = Style::new().bold());
-        for p in failures {
-            println!("{}", p.name());
-        }
-    }
-}
-
 macro_rules! error {
     ($($arg:tt)*) => {{
         $crate::logging::logln(

--- a/src/main.rs
+++ b/src/main.rs
@@ -13,6 +13,7 @@ use crate::finder::Finder;
 use crate::logging::init_logging;
 use crate::util::Options;
 use clap::Parser;
+use std::process::ExitCode;
 
 /// Operate on each project in a directory
 #[derive(Clone, Debug, Eq, Parser, PartialEq)]
@@ -28,14 +29,20 @@ struct Arguments {
     command: Command,
 }
 
-fn main() -> anyhow::Result<()> {
+fn main() -> ExitCode {
     let Arguments {
         opts,
         finder,
         command,
     } = Arguments::parse();
     init_logging(opts.verbosity());
-    let projects = finder.findall()?;
+    let projects = match finder.findall() {
+        Ok(projects) => projects,
+        Err(e) => {
+            error!("{e}"); // TODO: Display more
+            return ExitCode::FAILURE;
+        }
+    };
     command.run(opts, projects)
 }
 

--- a/src/main.rs
+++ b/src/main.rs
@@ -10,7 +10,7 @@ mod project;
 mod util;
 use crate::commands::Command;
 use crate::finder::Finder;
-use crate::logging::init_logging;
+use crate::logging::{init_logging, logerror};
 use crate::util::Options;
 use clap::Parser;
 use std::process::ExitCode;
@@ -39,7 +39,7 @@ fn main() -> ExitCode {
     let projects = match finder.findall() {
         Ok(projects) => projects,
         Err(e) => {
-            error!("{e}"); // TODO: Display more
+            logerror(e.context("Failed to list projects"));
             return ExitCode::FAILURE;
         }
     };

--- a/src/project/mod.rs
+++ b/src/project/mod.rs
@@ -1,6 +1,6 @@
 mod lang;
 pub(crate) use self::lang::*;
-use crate::cmd::{CommandError, CommandKind, CommandOutputError, CommandPlus};
+use crate::cmd::{CommandError, CommandKind, CommandPlus};
 use crate::util::get_ghrepo;
 use anyhow::Context;
 use cargo_metadata::{MetadataCommand, TargetKind};
@@ -177,7 +177,7 @@ impl Project {
         let r = self.readcmd("git", ["rev-parse", "--verify", "--quiet", "refs/stash"]);
         match r {
             Ok(stdout) => Ok(!stdout.is_empty()),
-            Err(CommandOutputError::Exit { rc, .. }) if rc.code() == Some(1) => Ok(false),
+            Err(CommandError::Exit { rc, .. }) if rc.code() == Some(1) => Ok(false),
             Err(e) => Err(e.into()),
         }
     }
@@ -206,7 +206,7 @@ impl Project {
         cmd
     }
 
-    pub(crate) fn readcmd<S, I>(&self, cmd: S, args: I) -> Result<String, CommandOutputError>
+    pub(crate) fn readcmd<S, I>(&self, cmd: S, args: I) -> Result<String, CommandError>
     where
         S: AsRef<OsStr>,
         I: IntoIterator<Item: AsRef<OsStr>>,

--- a/src/project/mod.rs
+++ b/src/project/mod.rs
@@ -194,7 +194,7 @@ impl Project {
             .stderr(Stdio::null())
             .run();
         match r {
-            Ok(_) => Ok(true),
+            Ok(()) => Ok(true),
             Err(CommandError::Exit { .. }) => Ok(false),
             Err(e) => Err(e.into()),
         }

--- a/src/util.rs
+++ b/src/util.rs
@@ -58,11 +58,10 @@ pub(crate) struct Runner {
 }
 
 impl Runner {
-    pub(crate) fn run(&self, p: &Project, opts: Options) -> Result<bool, CommandError> {
+    pub(crate) fn run(&self, p: &Project) -> Result<(), CommandError> {
         p.runcmd(&self.command)
             .args(self.args.iter())
             .kind(CommandKind::Run)
-            .keep_going(opts.keep_going)
             .run()
     }
 }

--- a/src/util.rs
+++ b/src/util.rs
@@ -1,4 +1,4 @@
-use crate::cmd::{CommandError, CommandKind, CommandOutputError, CommandPlus};
+use crate::cmd::{CommandError, CommandKind, CommandPlus};
 use crate::logging::Verbosity;
 use crate::project::Project;
 use clap::{ArgAction, Args};
@@ -116,8 +116,8 @@ pub(crate) fn get_ghrepo(p: &Path) -> anyhow::Result<Option<GHRepo>> {
         .check_output()
     {
         Ok(s) => Ok(GHRepo::from_url(s.trim()).ok()),
-        Err(CommandOutputError::Decode { .. }) => Ok(None),
-        Err(CommandOutputError::Exit { rc, .. }) if rc.code() == Some(2) => Ok(None),
+        Err(CommandError::Decode { .. }) => Ok(None),
+        Err(CommandError::Exit { rc, .. }) if rc.code() == Some(2) => Ok(None),
         Err(e) => Err(e.into()),
     }
 }

